src/runner/main.cpp

```diff
@@ -23,13 +23,78 @@
 
 #include <winrt/Windows.System.h>
 
+#include <Psapi.h>
+#include <RestartManager.h>
+
 #if _DEBUG && _WIN64
 #include "unhandled_exception_handler.h"
 #endif
 #include <common/notifications/fancyzones_notifications.h>
 
 extern "C" IMAGE_DOS_HEADER __ImageBase;
 
+const wchar_t ExplorerProcessName[] = L"explorer.exe";
+const wchar_t PTUpdateMessageBoxTitle[] = L"PowerToys";
+const wchar_t PTUpdateMessageBoxText[] = L"PowerToys was updated and some components require Windows Explorer to restart. Do you want to restart Windows Explorer now?";
+
+void GetProcessInfo(const std::wstring& processName, std::vector<RM_UNIQUE_PROCESS>& processInfos)
+{
+    DWORD bytesReturned{};
+    std::vector<DWORD> processIds{};
+    processIds.resize(1024);
+    DWORD processIdSize{ (DWORD)processIds.size() * sizeof(DWORD) };
+    EnumProcesses(processIds.data(), processIdSize, &bytesReturned);
+    while (bytesReturned == processIdSize)
+    {
+        processIdSize += processIdSize;
+        processIds.resize(processIdSize / sizeof(DWORD));
+        EnumProcesses(processIds.data(), processIdSize, &bytesReturned);
+    }
+    for (const DWORD& processId : processIds)
+    {
+        HANDLE hProcess = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, processId);
+        if (hProcess)
+        {
+            wchar_t name[MAX_PATH];
+            if (GetProcessImageFileName(hProcess, name, MAX_PATH) > 0)
+            {
+                if (wcscmp(processName.c_str(), PathFindFileName(name)) == 0)
+                {
+                    FILETIME creationTime{};
+                    FILETIME exitTime{};
+                    FILETIME kernelTime{};
+                    FILETIME userTime{};
+                    if (GetProcessTimes(hProcess, &creationTime, &exitTime, &kernelTime, &userTime))
+                    {
+                        processInfos.push_back({ processId, creationTime });
+                    }
+                }
+            }
+            CloseHandle(hProcess);
+        }
+    }
+}
+
+void RestartProcess(const std::wstring& processName)
+{
+    DWORD sessionHandle{};
+    WCHAR sessionKey[CCH_RM_SESSION_KEY + 1];
+    if (RmStartSession(&sessionHandle, 0, sessionKey) != ERROR_SUCCESS)
+    {
+        return;
+    }
+    std::vector<RM_UNIQUE_PROCESS> pInfo{};
+    GetProcessInfo(processName, pInfo);
+    if (pInfo.empty() ||
+        RmRegisterResources(sessionHandle, 0, nullptr, 1, pInfo.data(), 0, nullptr) != ERROR_SUCCESS)

```

Comment:
I think the third parameter should be `size(pInfo)` instead of `1`, since you can have multiple explorer.exe instances.

Reply:
Great catch! Changed.